doc/developer/design/20210401_volatile_sources.md

```diff
@@ -0,0 +1,294 @@
+# Volatile sources
+
+## Summary
+
+Our users want to connect Materialize to upstream sources that do not provide
+guarantees about what data they will present. These sources are in tension with
+the goal of providing a deterministic, replayable system.
+
+This design doc proposes to:
+
+  * Introduce "volatile source" as a term of art that describes a source that
+    does not have upstream durability.
+
+  * Use the terminology consistently in the docs site and error messages to
+    educate our users about volatile sources.
+
+  * Restrict the use of volatile sources in contexts where they could impede
+    future improvements to the system's determinism and correctness.
+
+## Goals
+
+The only goal is to **stabilize the [PubNub] and [SSE] sources**.
+
+Both of these are volatile sources with some worrying correctness properties,
+but their existence will massively improve the "wow factor" for demos of
+Materialize Cloud.
+
+[PubNub]: https://github.com/MaterializeInc/materialize/pull/6189
+[SSE]: https://github.com/MaterializeInc/materialize/pull/6208
+
+## Description
+
+### Terminology
+
+A "volatile source" is any source that cannot reliabily present the same data
+with the same timestamps across different instantiations.
+
+Examples of volatile sources:
+
+  * Any Kinesis source. Since Kinesis streams have mandatory retention periods
+    (by default 24 hours), a new instantiation of the source will not see
+    any data that has expired.
+  * A PubNub source, since new instantiations always start from the tail of the
+    stream, at least in the current implementation.
+  * A Kafka source with a non-infinite retention policy.
+
+Examples of nonvolatile sources:
+
+  * A file source, when the file is treated as append only.
+  * An S3 source, when the bucket is treated as append only.
+  * An non-upsert Kafka source, when the topic has infinite retention and no
+    compaction policy.
+  * An upsert Kafka source, when the topic has infinite retention but *may*
+    have a key compaction policy.
+
+It's important to note that volatility is a property of how a source is *used*,
+rather than fundamental to a source type. Kinesis is the rare exception. All
```

Comment:
I just want to make sure I understand the distinction correctly -- PubNub would be always volatile as well right? Unless you tried to write down the data and in so doing make it durable which would also work for Kinesis?

Reply:
Ah, yeah, I left out an important detail which is that PubNub actually supports storage and retrieval (https://www.pubnub.com/blog/introduction-to-storage-playback-on-the-pubnub-real-time-network/), just not for the demo streams.